harness/determined/tensorboard/metric_writers/callback.py

```diff
@@ -33,14 +33,25 @@ def on_train_step_end(
         step_id: int,
         num_batches: int,
         total_batches_processed: int,
-        metrics: List[Dict[str, Any]],
+        metrics: Dict[str, Any],
     ) -> None:
         if step_id <= 0:
             raise AssertionError(f"Expected step_id to be a positive int, but it is {step_id}")
-        for batch_idx, batch_metrics in enumerate(metrics):
+        metrics_seen = set()
+
+        # Log all batch metrics.
+        for batch_idx, batch_metrics in enumerate(metrics["batch_metrics"]):
             batches_seen = total_batches_processed + batch_idx
             for name, value in batch_metrics.items():
                 self._maybe_write_metric(name, value, batches_seen)
+                metrics_seen.add(name)
+
+        # Log avg metrics which were calculated by a custom reducer and are not in batch metrics.
+        batches_seen = total_batches_processed + num_batches
```

Comment:
question: is there anything wrong with just using the previous batches_seen value?

Reply:
Yes, there's not a strong guarantee that the previous loop was non-empty.  In practice that would probably work, but this is more robust because it reduces the number of assumptions required to run properly.